### PR TITLE
op-node: get block with sync status, make finalization more safe

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

This PR prepares the op-node derivation/driver bits for the op-proposer cleanup changes by securing the L2 finality better.

Changes:
- Method on driver to get a L2 block by number and sync status at the same time, so we have consistent data.
- Secure finalization to not allow inconsistent signals to affect the L2 finality status
- Add a `CurrentL1Finalized` to sync status to identify the L1 block we do recognize for derivation, matching the `CurrentL1` field which is last used for deriving the safe part of the L2 chain. `L1Finalized` in sync status just refers to the latest signal.
- Keep the engine queue progress in sync with the previous stage - the sync status / metrics weren't updating.
- When the L2 finality status changes, mark the engine as needing a forkchoice update, so we see geth match the finality status without waiting for blocks.
- Some miner batcher action-test 

**Tests**

Added an action test for L2 finalization edge cases, covering multiple batches and the edge case of a bad finality signal.

**Metadata**

Fix ENG-2772

### CODE REVIEW

**Comment by trianglesphere** [comment]
bedrock go tests are failing.

**Comment by protolambda** [comment]
It's one unrelated action test that's failing, due to a bad nonce for a transaction. I think the pending-nonce retrieval in geth doesn't interact so well with fast modeled block building. Will re-run the test and see if it passes.

File: op-node/rollup/driver/state.go [hunk 1]
```diff
@@ -30,8 +30,8 @@ type Driver struct {
 	// When the derivation pipeline is waiting for new data to do anything
 	idleDerivation bool
 
-	// Requests for sync status. Synchronized with event loop to avoid reading an inconsistent sync status.
-	syncStatusReq chan chan eth.SyncStatus
+	// Requests for synchronized event loop execution to avoid reading an inconsistent state
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:33]: Better comment would be `Requests to block the event loop`. There's no execution inside this channel, but it's instead holding the event loop for external execution (aka a mutex)

File: op-node/rollup/driver/state.go [hunk 2]
```diff
@@ -381,18 +373,41 @@ func (s *Driver) ResetDerivationPipeline(ctx context.Context) error {
 	}
 }
 
+func (s *Driver) syncStatus() *eth.SyncStatus {
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:376]: doc comment - particularly around invariants like should the event loop be blocked

File: op-node/rollup/driver/state.go [hunk 3]
```diff
@@ -381,18 +373,41 @@ func (s *Driver) ResetDerivationPipeline(ctx context.Context) error {
...
+		HeadL1:             s.l1State.L1Head(),
+		SafeL1:             s.l1State.L1Safe(),
+		FinalizedL1:        s.l1State.L1Finalized(),
+		UnsafeL2:           s.derivation.UnsafeL2Head(),
+		SafeL2:             s.derivation.SafeL2Head(),
+		FinalizedL2:        s.derivation.Finalized(),
+	}
+}
+
 func (s *Driver) SyncStatus(ctx context.Context) (*eth.SyncStatus, error) {
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:389]: doc comment

File: op-node/rollup/driver/state.go [hunk 4]
```diff
@@ -381,18 +373,41 @@ func (s *Driver) ResetDerivationPipeline(ctx context.Context) error {
...
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case resp := <-respCh:
-			return &resp, nil
-		}
+	}
+}
+
+func (s *Driver) BlockRefWithStatus(ctx context.Context, num uint64) (eth.L2BlockRef, *eth.SyncStatus, error) {
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:401]: doc comment

File: op-node/rollup/driver/state.go [hunk 5]
```diff
@@ -381,18 +373,41 @@ func (s *Driver) ResetDerivationPipeline(ctx context.Context) error {
...
+	wait := make(chan struct{})
+	select {
+	case s.stateReq <- wait:
+		resp := s.syncStatus()
+		ref, err := s.l2.L2BlockRefByNumber(ctx, num)
```
Comment (trianglesphere) [inline op-node/rollup/driver/state.go:406]: Why do we need to run block by number through the event loop?

File: op-e2e/actions/l1_replica.go [hunk 6]
```diff
@@ -177,8 +177,16 @@ func (s *L1Replica) L1Client(t Testing, cfg *rollup.Config) *sources.L1Client {
 // ActL1FinalizeNext finalizes the next block, which must be marked as safe before doing so (see ActL1SafeNext).
 func (s *L1Replica) ActL1FinalizeNext(t Testing) {
 	safe := s.l1Chain.CurrentSafeBlock()
-	finalizedNum := s.l1Chain.CurrentFinalizedBlock().NumberU64()
-	if safe.NumberU64() <= finalizedNum {
+	safeNum := uint64(0)
```
Comment (trianglesphere) [inline op-e2e/actions/l1_replica.go:180]: More a general observation that specific to this PR, I've been seeing a lot of churn the supporting/driver code

File: op-e2e/actions/l2_batcher_test.go [hunk 7]
```diff
@@ -87,3 +89,108 @@ func TestBatcher(gt *testing.T) {
 	require.False(t, isPending)
 	require.NotNil(t, vTx)
 }
+
+func TestL2Finalization(gt *testing.T) {
```
Comment (trianglesphere) [inline op-e2e/actions/l2_batcher_test.go:93]: I see that this test depends upon some of the changes in this PR (type changes + won't run if I just bring the test to develop), but it does seem like a grab bag of changes to have `op-e2e` driver fixes, a new test, & `op-node` changes bundled into one.

**Review by mslipper — APPROVED** [review]

File: op-node/rollup/driver/state.go (hunk 5 above)
Comment (protolambda) [reply op-node/rollup/driver/state.go:406]: We want block-by-number to be consistent with the returned status - reorgs may affect by-number requests.

File: op-e2e/actions/l1_replica.go (hunk 6 above)
Comment (protolambda) [reply op-e2e/actions/l1_replica.go:180]: New test framework, new churn. I'll try to improve it over time.

File: op-e2e/actions/l2_batcher_test.go (hunk 7 above)
Comment (protolambda) [reply op-e2e/actions/l2_batcher_test.go:93]: The proposer changes required a more consistent view of the chain, with accurate view handling bad finalization and reorgs between requests. So I split that up in a separate PR + wrote testing for it. It's not pretty, but small enough of a PR + includes testing.

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**